tests/devices/test_default_mixed_tf.py

```diff
@@ -0,0 +1,596 @@
+
+        state = dev.state
+        expected = np.zeros((4, 4))
+        expected[0, 0] = 1
+        assert np.allclose(state, expected, atol=tol, rtol=0)
+
+        @qml.qnode(dev, interface="tf", diff_method="backprop")
+        def circuit(a):
+            qml.Hadamard(wires=0)
+            qml.RZ(a, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        circuit(tf.constant(np.pi / 4))
+        state = dev.state
+
+        amplitude = np.exp(-1j * np.pi / 4) / 2
+        expected = np.array(
+            [[0.5, 0, amplitude, 0], [0, 0, 0, 0], [np.conj(amplitude), 0, 0.5, 0], [0, 0, 0, 0]]
+        )
+
+        assert np.allclose(state, expected, atol=tol, rtol=0)
+
+
+class TestDtypePreserved:
+    """Test that the user-defined dtype of the device is preserved for QNode
+    evaluation"""
+
+    @pytest.mark.parametrize("r_dtype", [np.float32, np.float64])
+    @pytest.mark.parametrize(
+        "measurement",
+        [
+            qml.expval(qml.PauliY(0)),
+            qml.var(qml.PauliY(0)),
+            qml.probs(wires=[1]),
+            qml.probs(wires=[2, 0]),
+        ],
+    )
+    def test_real_dtype(self, r_dtype, measurement, tol):
+        """Test that the user-defined dtype of the device is preserved
+        for QNodes with real-valued outputs"""
+        p = tf.constant(0.543)
+
+        dev = qml.device("default.mixed", wires=3, r_dtype=r_dtype)
+
+        @qml.qnode(dev, interface="tf", diff_method="backprop")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.apply(measurement)
+
+        res = circuit(p)
+        assert res.dtype == r_dtype
+
+    @pytest.mark.parametrize("c_dtype", [np.complex64, np.complex128])
+    @pytest.mark.parametrize(
+        "measurement",
+        [qml.state(), qml.density_matrix(wires=[1]), qml.density_matrix(wires=[2, 0])],
+    )
+    def test_complex_dtype(self, c_dtype, measurement, tol):
+        """Test that the user-defined dtype of the device is preserved
+        for QNodes with complex-valued outputs"""
+        p = tf.constant(0.543)
+
+        dev = qml.device("default.mixed", wires=3, c_dtype=c_dtype)
+
+        @qml.qnode(dev, interface="tf", diff_method="backprop")
+        def circuit(x):
+            qml.RX(x, wires=0)
+            return qml.apply(measurement)
+
+        res = circuit(p)
+        assert res.dtype == c_dtype
+
+
+class TestOps:
+    """Unit tests for operations supported by the default.qubit.autograd device"""
+
+    def test_multirz_jacobian(self):
+        """Test that the patched numpy functions are used for the MultiRZ
+        operation and the jacobian can be computed."""
+        wires = 4
+        dev = qml.device("default.mixed", wires=wires)
+
+        @qml.qnode(dev, interface="tf", diff_method="backprop")
+        def circuit(param):
+            qml.MultiRZ(param, wires=[0, 1])
+            return qml.probs(wires=list(range(wires)))
+
+        param = tf.Variable(0.3, trainable=True)
```

Comment:
Note: we usually don't explicitly set the trainable attribute of `tf.Variable` objects in PennyLane because Variables are trainable by default when used inside a `GradientTape` context (personally this has been my first time coming across this attribute, just checked with Josh on this :smile: ).

Reply:
Really 😄? I see `trainable=True` explicitly set in almost every test and example. If they're trainable by default then I guess it's a stylistic choice.

Reply:
Yeah likely